src/verifier/step3/Step3Logic.sol

```diff
@@ -0,0 +1,359 @@
+        }
+        counter++;
+
+        // Abosrb u
+        elementsToHash[counter] = U1.u;
+        counter++;
+
+        // Absorb X
+        for (uint256 i = 0; i < U1.X.length; i++) {
+            uint256 entry = U1.X[i];
+            uint256 limb1 = (0x000000000000000000000000000000000000000000000000ffffffffffffffff & entry);
+            uint256 limb2 = (0x00000000000000000000000000000000ffffffffffffffff0000000000000000 & entry) >> 64;
+            uint256 limb3 = (0x0000000000000000ffffffffffffffff00000000000000000000000000000000 & entry) >> 128;
+            uint256 limb4 = (0xffffffffffffffff000000000000000000000000000000000000000000000000 & entry) >> 192;
+
+            elementsToHash[counter] = limb1;
+            counter++;
+            elementsToHash[counter] = limb2;
+            counter++;
+            elementsToHash[counter] = limb3;
+            counter++;
+            elementsToHash[counter] = limb4;
+            counter++;
+        }
+
+        //U2.absorb_in_ro
+        // Absorb comm_W
+        elementsToHash[counter] = U2.comm_W.x;
+        counter++;
+        elementsToHash[counter] = U2.comm_W.y;
+        counter++;
+        if (Pallas.isInfinity(U2.comm_W)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        // Absorb X
+        for (uint256 i = 0; i < U2.X.length; i++) {
+            elementsToHash[counter] = U2.X[i];
+            counter++;
+        }
+
+        // Absorb comm_T
+        Pallas.PallasAffinePoint memory comm_T = Pallas.decompress(nifs.compressed_comm_T);
+        elementsToHash[counter] = comm_T.x;
+        counter++;
+        elementsToHash[counter] = comm_T.y;
+        counter++;
+        if (Pallas.isInfinity(comm_T)) {
+            elementsToHash[counter] = 1;
+        }
+        else {
+            elementsToHash[counter] = 0;
+        }
+        counter++;
+
+        require(counter == NUM_FE_FOR_RO); 
+
+        // uint32 absorbLen = uint32(counter);
+        // uint32 squeezeLen = 1;
+        // uint32 domainSeparator = 0;
+
+        SpongeOpLib.SpongeOp memory absorb = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Absorb, uint32(counter));
+        SpongeOpLib.SpongeOp memory squeeze = SpongeOpLib.SpongeOp(SpongeOpLib.SpongeOpType.Squeeze, 1);
+        SpongeOpLib.SpongeOp[] memory pattern = new SpongeOpLib.SpongeOp[](2);
+        pattern[0] = absorb;
+        pattern[1] = squeeze;
+        IOPatternLib.IOPattern memory p = IOPatternLib.IOPattern(pattern);
+        
+        NovaSpongePallasLib.SpongeU24Pallas memory sponge = NovaSpongePallasLib.start(p, 0);
+
+        sponge = NovaSpongePallasLib.absorb(sponge, elementsToHash);
+
+        (, uint256[] memory output) = NovaSpongePallasLib.squeeze(sponge, 1);
+        sponge = NovaSpongePallasLib.finishNoFinalIOCounterCheck(sponge);
+
+        // uint256 r = output[0] & 0x00000000000000000000000000000000ffffffffffffffffffffffffffffffff;
```

Comment:
redundant comment?

Reply:
Agreed, these comments were made when I was minimizing the number of local variables because of a stack too large issue.